### ISSUE TITLE
Allow assets to go through the processing pipeline as well

### ISSUE THREAD

**Issue by mitar** [meteor/meteor#5600]
Currently, assets are added "as is". But it would be great if you could mark them to be processed as well. For example, transpile ES2015 into JS so that it can then be included into a web worker.

**Comment by mitar** [comment]
(Additional idea would be generating a web font automatically.)

**Comment by stubailo** [comment]
Actually this is already possible - you can have a build plugin that adds an asset, as long as your source files have some extension other than `.js`. By definition, files in `public` and `private` are _not_ processed by any plugins.

For example, you could have an extension called `.worker.js` that is processed by your new build plugin which just applies the same transformation as `ecmascript` and then calls `addAsset` instead of `addJavaScript`.

So, in conclusion, in the Meteor build system, files that are processed are by definition source files and not assets; however source files can be compiled into assets, which would meet your use case.

In an app, it is not possible to use the same file as a source and asset, but this is possible in a package.

**Comment by mitar** [comment]
Thanks!

**Comment by mitar** [comment]
Hmmm. But I cannot combine multiple files into one output asset file, no?

**Comment by stubailo** [comment]
I think you can, in much the same way the LESS compiler compiles multiple LESS files into one CSS file. Right now the documentation on this is lacking, but we're planning on fixing that in the Guide article about making packages. The best places to look for examples are:

1. The class you would need to subclass: https://github.com/meteor/meteor/blob/devel/packages/caching-compiler/multi-file-caching-compiler.js
2. An example where LESS uses it: https://github.com/meteor/meteor/blob/devel/packages/less/plugin/compile-less.js

**Comment by mitar** [comment]
Hm, so what I would ideally want is:

```javascript
api.addAssets([
  'file1.worker.js',
  'file2.worker.js'
], {output: 'worker.js'});
```

Or something?

It would be great if that would be possible even for normal JavaScript, to say, make me this asset but combining these files and minimizing them in production, and store it under `worker.js` as an asset.

**Comment by stubailo** [comment]
Yeah I mean that's a pretty hard limitation of the current build system, that there isn't a sane way to configure build plugins like that. The best way is to also register a file like `my-plugin-config.json` and put the configuration in there, but that just makes the logic more complicated. Sorry there isn't a better story for this; I bet the Webpack people have something...

**Comment by stubailo** [comment]
Actually, there might be a way to get `fileOptions` inside the build plugin - have you looked into that?

**Comment by mitar** [comment]
Not yet. :-)

**Comment by stubailo** [comment]
It looks like the official API doesn't support it; perhaps we can have a simple patch that passes through `fileOptions`?

**Comment by mitar** [comment]
So then the hacky way would be: create a plugin for custom extensions, have a custom `fileOptions`, use caching plugin to combine all files for the same file option together, and then output it at the end.

I am only not sure if `MultiFileCachingCompiler` support behavior like "output at the end when all `addFiles` calls have been made", but only "output when all files referenced by the file have been made"?

**Comment by stubailo** [comment]
The new build plugins only compile files in the context of the app; that is, you can have access to all files added with this file extension by any packages in the current build process. I'm not sure if that covers your use case or not?

**Comment by mitar** [comment]
So, I would like to make a package which combines multiple source file into one output JS worker file which then client-side can use as an asset/worker content.

But yea, probably if I can have access to all files with an extension can help, just the question is, can I run something at the end of the whole build system, when I know which all files have been added and with which options?

**Comment by stubailo** [comment]
Yes. The plugins always run at the time that you desire, when everything is known. So you can actually combine files across different packages if you want to.

**Comment by stubailo** [comment]
This is one of the primary features of the new build system. In Meteor 1.1 and below, packages were each built separately one at a time. In Meteor 1.2, all of the packages and the app are built at once, in one step.

**Comment by mitar** [comment]
And there is still not guide. :-) OK, let's try to reverse engineer this.

**Comment by mitar** [comment]
Ah, but `fileOptions` should be passed to plugins, this is still missing?

**Comment by stubailo** [comment]
Oh, I guess I'm wrong, apparently you can get fileOptions: https://github.com/meteor/meteor/blob/d9b895e6ec02dc46cf40a9d9d798dbdd77a2fc2a/packages/babel-compiler/babel-compiler.js#L21

So perhaps what you want is already possible! (Except not in an app, since there is no way to pass options there...)

**Comment by mitar** [comment]
Thanks, I will have to look into this more.

**Comment by stubailo** [comment]
OK, I hope I answered your questions well enough for you to get something done!
